File: iosApp/iosApp/ConsentManager.swift
//
//  ConsentManager.swift
//  Block Blast
//
//  Wraps Google's User Messaging Platform (UMP) so GDPR / regional consent
//  is gathered before any AdMob request fires. Google requires that the
//  AdMob SDK be initialised *after* consent, so `MobileAds.shared.start`
//  is invoked from inside this class once `canRequestAds` is true.
//
//  Add the SDK via Xcode → File → Add Package Dependencies:
//      https://github.com/googleads/swift-package-manager-google-user-messaging-platform.git
//
//  Info.plist must include a valid `GADApplicationIdentifier`. Consent
//  debug/testing overrides can be wired inside `gather(from:onReadyToRequestAds:)`
//  via `UMPDebugSettings`.
//

import Foundation
import UIKit
import GoogleMobileAds
import UserMessagingPlatform
import AppTrackingTransparency

@MainActor
final class ConsentManager {
    static let shared = ConsentManager()

    private var mobileAdsInitialized = false
    private var readyFired = false
    private var gatherStarted = false
    private var attFlowStarted = false
    private var didBecomeActiveObserver: NSObjectProtocol?
    private var attRetryCount = 0

    private init() {}

    /// Whether AdMob requests are currently permitted.
    /// UMP caches consent across launches, so this may return `true`
    /// immediately on relaunch even before `gather` completes.
    var canRequestAds: Bool {
        ConsentInformation.shared.canRequestAds
    }

    /// Request ATT first, then request UMP consent info, show the form if
    /// required, and start the AdMob SDK only after both gates complete. Call
    /// after the app UI is visible, for example from the root controller's
    /// first `viewDidAppear`.
    ///
    /// `onReadyToRequestAds` fires exactly once per process — on the first
    /// transition to "can request ads".
    func gather(
        from rootViewController: UIViewController?,
        onReadyToRequestAds: @escaping () -> Void
    ) {
        guard !gatherStarted else { return }
        gatherStarted = true

        requestATTIfNeeded(from: rootViewController) { [weak self, weak rootViewController] in
            guard let self else { return }
            self.gatherUMPConsent(from: rootViewController, onReadyToRequestAds: onReadyToRequestAds)
        }
    }

    private func gatherUMPConsent(
        from rootViewController: UIViewController?,
        onReadyToRequestAds: @escaping () -> Void
    ) {
        let parameters = RequestParameters()
        // Debug overrides (EEA geography, test device hashes) can be set here
        // while testing localized consent UI.
        //
        // let debug = DebugSettings()
        // debug.geography = .EEA
        // debug.testDeviceIdentifiers = ["TEST-DEVICE-HASH"]
        // parameters.debugSettings = debug

        ConsentInformation.shared.requestConsentInfoUpdate(
            with: parameters
        ) { [weak self] requestError in
            guard let self else { return }
            if let requestError {
                print("[ConsentManager] info update failed: \(requestError.localizedDescription)")
                if self.canRequestAds {
                    self.fireReady(onReadyToRequestAds)
                }
                return
            }

            guard let presenter = rootViewController else {
                if self.canRequestAds {
                    self.fireReady(onReadyToRequestAds)
                }
                return
            }

            ConsentForm.loadAndPresentIfRequired(from: presenter) { [weak self] formError in
                guard let self else { return }
                if let formError {
                    print("[ConsentManager] form error: \(formError.localizedDescription)")
                }
                if self.canRequestAds {
                    self.fireReady(onReadyToRequestAds)
                }
            }
        }

        // Already-cached consent path. ATT has completed before this method is
        // called, so ads may start as soon as UMP permits requests.
        if canRequestAds {
            fireReady(onReadyToRequestAds)
        }
    }

    private func requestATTIfNeeded(
        from presenter: UIViewController?,
        completion: @escaping () -> Void
    ) {
        guard !attFlowStarted else { return }
        attFlowStarted = true

        print("[ConsentManager] ATT flow requested. currentStatus=\(Self.attStatusDescription)")
        scheduleATTRequestAfterUISettles(from: presenter, completion: completion)
    }

    private func scheduleATTRequestAfterUISettles(
        from presenter: UIViewController?,
        completion: @escaping () -> Void
    ) {
        let delay: TimeInterval = attRetryCount == 0 ? 1.0 : 0.35
        DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak self, weak presenter] in
            guard let self else { return }

            guard self.isReadyToRequestATT() else {
                self.attRetryCount += 1
                print(
                    "[ConsentManager] ATT deferred. appState=\(UIApplication.shared.applicationState.rawValue), " +
                    "hasWindow=\(presenter?.view.window != nil), retry=\(self.attRetryCount)"
                )
                if self.attRetryCount < 8 {
                    let retry: () -> Void = { [weak self, weak presenter] in
                        guard let self else { return }
                        self.scheduleATTRequestAfterUISettles(from: presenter, completion: completion)
                    }
                    if UIApplication.shared.applicationState == .active {
                        retry()
                    } else {
                        self.runWhenAppBecomesActive(retry)
                    }
                } else {
                    print("[ConsentManager] ATT skipped because UI never became ready.")
                    completion()
                }
                return
            }

            self.performATTIfNeeded(completion: completion)
        }
    }

    private func performATTIfNeeded(completion: @escaping () -> Void) {
        let status = ATTrackingManager.trackingAuthorizationStatus
        print("[ConsentManager] ATT request point reached. status=\(Self.attStatusDescription)")

        guard status == .notDetermined else {
            completion()
            return
        }

        ATTrackingManager.requestTrackingAuthorization { status in
            DispatchQueue.main.async {
                print("[ConsentManager] ATT completed. status=\(Self.attStatusDescription(status))")
                completion()
            }
        }
    }

    private func isReadyToRequestATT() -> Bool {
        UIApplication.shared.applicationState == .active
    }

    private func runWhenAppBecomesActive(_ block: @escaping () -> Void) {
        guard didBecomeActiveObserver == nil else { return }
        didBecomeActiveObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let self else { return }
            if let observer = self.didBecomeActiveObserver {
                NotificationCenter.default.removeObserver(observer)
                self.didBecomeActiveObserver = nil
            }
            block()
        }
    }

    private static var attStatusDescription: String {
        attStatusDescription(ATTrackingManager.trackingAuthorizationStatus)
    }

    private static func attStatusDescription(_ status: ATTrackingManager.AuthorizationStatus) -> String {
        switch status {
        case .notDetermined:
            return "notDetermined"
        case .restricted:
            return "restricted"
        case .denied:
            return "denied"
        case .authorized:
            return "authorized"
        @unknown default:
            return "unknown"
        }
    }

    private func fireReady(_ onReadyToRequestAds: () -> Void) {
        if !mobileAdsInitialized {
            mobileAdsInitialized = true
            MobileAds.shared.start(completionHandler: nil)
        }
        if !readyFired {
            readyFired = true
            onReadyToRequestAds()
        }
    }
}

File: composeApp/src/androidMain/kotlin/ge/yet3/blokblast/ads/consent/ConsentManager.kt
package ge.yet3.blokblast.ads.consent

import android.app.Activity
import android.content.Context
import android.util.Log
import com.google.android.gms.ads.MobileAds
import com.google.android.ump.ConsentInformation
import com.google.android.ump.ConsentRequestParameters
import com.google.android.ump.UserMessagingPlatform
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Google User Messaging Platform (UMP) wrapper.
 *
 * Handles GDPR / regional consent before any AdMob request fires. Call
 * [gather] from the host Activity's `onCreate`. The [onReadyToRequestAds]
 * callback fires exactly once — either after the user finishes the consent
 * form or immediately if consent is not required (or was previously given).
 *
 * The AdMob SDK itself is initialised here, *after* consent, to comply with
 * Google's UMP requirements — initialising earlier risks firing requests
 * before a GDPR-required form has been shown.
 */
object ConsentManager {

    private const val TAG = "ConsentManager"

    private val mobileAdsInitialized = AtomicBoolean(false)
    private val readyFired = AtomicBoolean(false)

    @Volatile
    private var consentInformation: ConsentInformation? = null

    /**
     * Whether ads can currently be requested. Treat as `false` until
     * [gather] has reported ready at least once in this process.
     */
    fun canRequestAds(context: Context): Boolean {
        val info = consentInformation
            ?: UserMessagingPlatform.getConsentInformation(context).also { consentInformation = it }
        return info.canRequestAds()
    }

    /**
     * Request UMP consent info, show the form if required, and initialise the
     * AdMob SDK once the user can legally receive ads.
     *
     * Safe to call multiple times; the [onReadyToRequestAds] callback fires
     * only on the first transition to "can request ads" per process.
     */
    fun gather(activity: Activity, onReadyToRequestAds: () -> Unit) {
        val params = ConsentRequestParameters.Builder()
            // Debug settings can be attached here when testing in the EEA:
            //   .setConsentDebugSettings(ConsentDebugSettings.Builder(activity)
            //       .setDebugGeography(DEBUG_GEOGRAPHY_EEA)
            //       .addTestDeviceHashedId("TEST-DEVICE-HASH").build())
            .build()

        val info = UserMessagingPlatform.getConsentInformation(activity).also {
            consentInformation = it
        }

        info.requestConsentInfoUpdate(
            activity,
            params,
            {
                UserMessagingPlatform.loadAndShowConsentFormIfRequired(activity) { formError ->
                    if (formError != null) {
                        Log.w(
                            TAG,
                            "Consent form error ${formError.errorCode}: ${formError.message}",
                        )
                    }
                    if (info.canRequestAds()) fireReady(activity, onReadyToRequestAds)
                }
            },
            { requestError ->
                Log.w(
                    TAG,
                    "Consent info update failed ${requestError.errorCode}: ${requestError.message}",
                )
                // Network / config failure — fall back to whatever the cached
                // consent state allows. Users in non-GDPR regions still get ads.
                if (info.canRequestAds()) fireReady(activity, onReadyToRequestAds)
            },
        )

        // If we already have valid consent from a previous session, fire
        // immediately so the first interstitial can pre-load during the UMP
        // round-trip instead of waiting for it to return.
        if (info.canRequestAds()) fireReady(activity, onReadyToRequestAds)
    }

    private fun fireReady(context: Context, onReadyToRequestAds: () -> Unit) {
        if (mobileAdsInitialized.compareAndSet(false, true)) {
            MobileAds.initialize(context.applicationContext) { /* no-op */ }
        }
        if (readyFired.compareAndSet(false, true)) {
            onReadyToRequestAds()
        }
    }
}
